Support extension discovery suggestions in products list

haszari opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [x] Have you tested your feature works correctly in a range of browsers and devices?


<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Adds support for displaying subtle and appropriate in-product suggestion messaging for WooCommerce Marketplace extensions. These messages are driven by data from an API on WooCommerce.com.

### How to test the changes in this Pull Request:
1. Checkout this branch and get it set up as the WooCommerce plugin in your development environment.
2. Log in as a store administrator and view suggestions in the products admin screen – there are different suggestions displayed in the empty state and within the list of products.
3. Test dismissing suggestions. Dismissed suggestions are stored in user settings with key `wp_wc_marketplace_suggestions_dismissed_suggestions` (delete row to reset).

You can also test these changes on [this jurassic.ninja site](https://married-white-clawed.jurassic.ninja) – ping me (@haszari) for a user account to test.

### Screenshot

<img width="923" alt="screen_shot_2019-02-20_at_11 31 04_am-1" src="https://user-images.githubusercontent.com/4167300/53063003-87cf0980-3527-11e9-9787-69d59ca7c835.png">

rochow committed:
Update class-wc-product-data-store-cpt.php

Updating find_matching_product_variation() function to be faster

rochow committed:
Update class-wc-product-data-store-cpt.php

rochow committed:
Update class-wc-product-data-store-cpt.php

thenbrent committed:
Add WC_Tests_Webhook_Functions::create_webhook()

To make it reusable.

thenbrent committed:
Add $limit param to wc_load_webhooks()

And the corresponding data stores. Defaults to null, meaning do not
limit, for backward compatibility.

thenbrent committed:
Only load 100 webhooks per request

To avoid slowing down page loads on sites with a large numbers of
webhooks.

thenbrent committed:
Add $status param to wc_load_webhooks()

And the corresponding data stores. Defaults to '', meaning do not
load only webhooks with a specific status. This maintains backward
compatibility. However, the call to wc_load_webhooks() within
WooCommerce::load_webhooks() can now only load active webhooks, as
they are the only ones that should be enqueued.

thenbrent committed:
Only load active webhooks on each request

To avoid slowing down page loads on sites with a large number of
disabled or paused webhooks, which do not need to be loaded or
enqueued.

thenbrent committed:
Use search_webhooks()

To avoid duplicate SQL

thenbrent committed:
PHPCS fixes

Add fixes required for Travis to pass the PR, but which aren't
related to the PR diff.

Commit:
Test against $wp_filter instead of has_filter()

Because we don't have the same WC_Webhook instance as used in
wc_load_webhooks(), so it's impossible to check if the same
object's process() method is attached as a callback.

Commit:
Fix test_wc_load_webhooks_status_and_limit()

As the status is being explicitly passed to wc_load_webhooks(), it
will load webhooks with that status, not just active.

mikejolley committed:
Add adjust stock handling when deleting items

mikejolley committed:
Use adjust stock handling

mikejolley committed:
Revert "Add adjust stock option to order class when adding items"

This reverts commit 4a0447a1509087193d612815d528bf2d022b976e.

mikejolley committed:
Allow notes to be updated after other order screen events

mikejolley committed:
Revert "Add adjust stock handling when deleting items"

This reverts commit 0a5c77d93da6930be5b081f47f5e2a0bd33039e9.

mikejolley committed:
Stock handling in ajax class only

mikejolley committed:
Update notes with response

mikejolley committed:
wc_maybe_adjust_line_item_product_stock helper

mikejolley committed:
Use helper and return notes.

mikejolley committed:
Merge remote-tracking branch 'origin' into pr/22125

mikejolley committed:
Remove repetition from wc_get_customer_default_location and add validity check

mikejolley committed:
Whenn loading address fields, enforce valid address

mikejolley committed:
Set country value when loading fields rather than in template files

mikejolley committed:
Add missing placeholder

mikejolley committed:
PHPCS fixes

mikejolley committed:
Always round to precision. Negate need to call get_tax_total to handle rounding. Instead it just sums.

mikejolley committed:
Correct use of rounding functions in cart class

mikejolley committed:
Calculated items_subtotal_tax using rounded tax subtotals.

mikejolley committed:
set_item_discount_amounts should support woocommerce_tax_round_at_subtotal

mikejolley committed:
set_coupon_discount_amounts should support woocommerce_tax_round_at_subtotal

mikejolley committed:
update_taxes can safely round all values, as lines would already be rounded if applicable

mikejolley committed:
Make total recalc use unrounded tax values to fix test WC_Tests_Order_Coupons::test_add_coupon_to_order

mikejolley committed:
Order items respect wc_round_tax_total

mikejolley committed:
Merge branch 'fix/21871' into refactor/tax-rounding-methods

mikejolley committed:
Code style improvements

mikejolley committed:
Query tweaks

mikejolley committed:
Exit early, add back fallback code

mikejolley committed:
Removed unused query

thenbrent committed:
Remove unused $wpdb

thenbrent committed:
Clear webhook transients when status changes

Background discussion: https://github.com/woocommerce/woocommerce/pull/21427#discussion_r246881785

mikejolley committed:
phpcs

mikejolley committed:
Trigger loaded hook once plugins are loaded.

peterfabian committed:
Merge branch 'master' into fix/20937-v0.1

peterfabian committed:
Updated check to only cover non-legacy REST API requests.
Updated filter name.

peterfabian committed:
Renamed function

As legacy REST API will be removed soon anyway.

peterfabian committed:
Updated todo to use the common format.

rrennick committed:
don't force shop loop display mode to products when query page >1

rrennick committed:
phpcs sniff fixes for wc-template-functions.php

rrennick committed:
phpcs sniff fixes for class-wc-breadcrumb.php

rrennick committed:
two more sniffs in wc-template-functions.php

mikejolley committed:
Dump max width in notices

mikejolley committed:
Added woocommerce_paypal_force_one_line_item filter

rrennick committed:
limit page 2 not showing products to shop showing subcategories

mikejolley committed:
novalidate order review form

mikejolley committed:
Switch span to paragraph

mikejolley committed:
BlockUI on order form submit

Commit:
Add filter woocommerce_update_product_stock_query

Commit:
Ignore WordPress.DB.PreparedSQL.NotPrepared

peterfabian committed:
Added check for POST variable existance.

mikejolley committed:
phpcs

claudiulodro committed:
Add aria label for shop orderby

mikejolley committed:
Use mb functions for limiting string length

mikejolley committed:
Tests

Tofandel committed:
The "for" attribute of a label for a radio input is invalid

The main label of a radio supposedly targeting the first value of the radios contains only the value instead of theId_theValue which is the id of the radio

mikejolley committed:
Update CA address format

kloon committed:
Custom payment sections not loading fields due to logic error.

kloon committed:
Use name instead of singular_name for archive pages breadcrumbs. singular_name should only be used on single pages.

kloon committed:
PHPCS fixes

rrennick committed:
update product table to correctly calculate net product sales

haszari committed:
in-product discovery - render container for product-list suggestions

haszari committed:
in-product discovery - render suggestions in products list

yratof committed:
Added filter for product title classes within product loop

mikejolley committed:
Merge pull request #22681 from woocommerce/fix/22678

Added check for POST variable existence.

mikejolley committed:
Merge pull request #22664 from woocommerce/fix/21788

BlockUI on order form submit

mikejolley committed:
Merge pull request #22663 from woocommerce/fix/22559

Switch span to paragraph in user profile view

mikejolley committed:
Merge pull request #22653 from woocommerce/update/22568

Added `woocommerce_paypal_force_one_line_item` filter

mikejolley committed:
Merge pull request #22652 from woocommerce/fix/22507

Dump `max-width` in notices

mikejolley committed:
Merge pull request #22420 from woocommerce/refactor/tax-rounding-methods

Refactor tax rounding methods for consistency

mikejolley committed:
Merge fix/21858

claudiosanches committed:
Load REST API endpoints before generate webhook payload

claudiosanches committed:
Fix special chars on webhook secret

By default some values are encoded before saved into the database, so it's required to decode those chars before generate the webhook signature.

Fixes #22500

haszari committed:
in-product discovery - don't promote extensions that are already installed:
- + move suggestions client code to a module, enqueued only on relevant pages

haszari committed:
register underscore dependency for marketplace script

haszari committed:
in-product discovery - show items if companion extension(s) installed

haszari committed:
in-product discovery - enqueue styles as separate module

haszari committed:
fix reference to data global (lint)

haszari committed:
don't use ES6 unnecessarily in gruntfile - breaks the linter/build

haszari committed:
remove use of ES6 features in marketplace suggestions (const, =>)

haszari committed:
style marketplace cards grid in empty products screen

haszari committed:
style marketplace suggestions

haszari committed:
fix issue with footer overlapping empty products layout - hide footer

haszari committed:
style marketplace suggestions list banner

mikejolley committed:
Merge pull request #22722 from woocommerce/fix/22500

Fix special chars on webhook secret

mikejolley committed:
Merge pull request #22721 from woocommerce/fix/webhooks-payload

Load REST API endpoints before generate webhook payload

yratof committed:
Updated product loop title filter with esc_attr()

mikejolley committed:
Merge pull request #22705 from woocommerce/fix/22699

Archive breadcrubs should use plural name not singular

mikejolley committed:
Merge pull request #22683 from woocommerce/fix/22679

Add aria label for shop orderby selector

mikejolley committed:
Merge pull request #22711 from woocommerce/fix/22656

Calculation issues on single product sales by product graph/report

mikejolley committed:
Merge pull request #22704 from woocommerce/fix/22696

Custom payment sections not loading fields due to logic error.

mikejolley committed:
Merge pull request #22672 from Chouby/update_stock_filter

Allow to filter the direct sql query to update the product stock

mikejolley committed:
Merge pull request #22717 from yratof/yratof-patch-1

Added filter for product title classes within product loop

mikejolley committed:
Merge pull request #22690 from Tofandel/patch-1

The "for" attribute of a label for a radio input is invalid

mikejolley committed:
Merge branch 'fix/21830'

mikejolley committed:
Check for _reduced_stock meta when restocking refunded items

mikejolley committed:
Refactor should_deliver logic into managable chunks

mikejolley committed:
Create is_valid_resource method to check for invalid statuses

mikejolley committed:
Avoid handling wc-auto-draft

WC was erroneously adding wc- prefix to the core WP auto-draft status.

#22380 registered it formally but we don't need it.

I've reverted #22380 and handled the prefix correctly.

mikejolley committed:
Update CRUD update hook based on status transition

mikejolley committed:
Ignore old pre-crud actions

mikejolley committed:
Update the tests

haszari committed:
use attributes & DOM api for table row banner render - html snippet is too flexible/risky

haszari committed:
use attributes + DOM (not html snippet) for products empty state suggestions

haszari committed:
ensure list suggestions render correctly with optional url/copy attributes

haszari committed:
styling products empty state OBW style, including existing add/import + suggestions

haszari committed:
fill out products empty state suggestions:
- add back product suggestions data
- hide page admin elements, focus layout on onboarding
- bulk-fix indentation in admin.scss
- reinstate default BlankState styling (overrides are intended for products only, accidentally overwrote)

haszari committed:
products blank state - OBW style suggestions (missed in previous commit)

haszari committed:
move hard-coded suggestion data to woo API (WCCOM api coming soon)

mikejolley committed:
Update photoswipe

mikejolley committed:
Escape html entities before passing to WP/photoswipe

renovate-bot committed:
Update dependency lint-staged to v8.1.4

mikejolley committed:
If date paid is not set but payment complete status has passed, set it anyway Ref: #22688

mikejolley committed:
Variations cannot be drafts - set to private.

mikejolley committed:
phpcs

mikejolley committed:
Fix save logic by moving setter into loop grabbing the value

Broken by https://github.com/woocommerce/woocommerce/pull/22650 phpcs changes

mikejolley committed:
Check for decoded taxonomy name when unsetting

mikejolley committed:
Revert debug code

peterfabian committed:
Put back status parameter after extra processing.

This allows correct further processing of $request, e.g. for next/previous links, etc.

haszari committed:
pull suggestion data from (proposed) S3 endpoint:
- leave hard-coded data in place until endpoint is available

haszari committed:
update styling of products table suggestion + add dismiss button

haszari committed:
ajax api for dismissing specific marketplace suggestions (first cut)

yratof committed:
Hard coded class now filterable

Ability to add library classes to price instead of overwriting template

renovate-bot committed:
Update dependency eslint to v5.14.0

chris-kreidl committed:
added IDs to other input fields for consistency

haszari committed:
ensure suggestion dismiss API / nonce is provided to js (missed in last commit)

haszari committed:
hide suggestions immediately after user clicks x (using jQuery.fadeOut)

haszari committed:
support dismissing empty-state suggestions + polish dismiss styling:\n- hover style for x\n- disable border/underline active state inherited from <a>\n- layout styling for empty-state suggestions, button on right\n- make test data more like mockup (no category recommendation in list)

mikejolley committed:
Update from master

mikejolley committed:
Select an option vs state

mikejolley committed:
Merge pull request #22758 from ckreidl/master

added IDs to other input fields for consistency

mikejolley committed:
Merge pull request #22752 from woocommerce/renovate/eslint-5.x

Update dependency eslint to v5.14.0

mikejolley committed:
Merge pull request #22734 from woocommerce/renovate/lint-staged-8.x

Update dependency lint-staged to v8.1.4

mikejolley committed:
update phpcs line

mikejolley committed:
Merge pull request #22536 from woocommerce/update/21524

Delay `woocommerce_loaded` hook until all plugins are loaded.

mikejolley committed:
Update reduced stock amount under some circumstances

mikejolley committed:
Set woocommerce_load_webhooks_limit to no limit

mikejolley committed:
phpcs

mikejolley committed:
Update versions

mikejolley committed:
version comment

mikejolley committed:
Removed limit option from get_webhooks_ids - since transient stores all ids anyway, this logic makes more sense elsewhere

mikejolley committed:
Add limit to wc_load_webhooks

mikejolley committed:
Avoid getting all webhook ids

mikejolley committed:
get_webhooks_ids escaping

mikejolley committed:
phpcs

mikejolley committed:
Dedicated count method

mikejolley committed:
Performance: Support pagination to avoid double queries to search webhooks

mikejolley committed:
Update from master

mikejolley committed:
Update from master

mikejolley committed:
Update from master

mikejolley committed:
Move BN partner ID

mikejolley committed:
Merge pull request #22741 from woocommerce/fix/22701

Put back status parameter after extra processing.

mikejolley committed:
wc_maybe_adjust_line_item_product_stock should return false if nothing happens

mikejolley committed:
Update from master

mikejolley committed:
Missed null check in wc_load_webhooks

mikejolley committed:
Merge pull request #22729 from woocommerce/fix/22489

Check for _reduced_stock meta when restocking refunded items

mikejolley committed:
Merge branch 'fix/21463'

mikejolley committed:
Merge branch 'master' of https://github.com/woocommerce/woocommerce

renovate-bot committed:
Update dependency eslint to v5.14.1

peterfabian committed:
Renamed filter and reused the function in auth code.

peterfabian committed:
Updated call to reflect function change.

claudiosanches committed:
Fixed coding standards

claudiosanches committed:
Catch WC_Data_Exception to avoid black screen

Also incldued a check to avoid duplicated messages about invalid billing
address emails.

claudiosanches committed:
Fixed coding standards

renovate-bot committed:
Update dependency mocha to v6

claudiosanches committed:
Use short cuts for wpdb properties

claudiosanches committed:
Merge branch 'master' into performance/find-matching-variations

claudiosanches committed:
Only disable WordPress.DB.SlowDBQuery.slow_db_query_tax_query

claudiosanches committed:
PHPCS ignore only what is necessary

haszari committed:
update products list empty state (incl. suggestions) to latest designs

peterfabian committed:
Perhaps it does not make much sense to filter empty URI if it's REST request or not.

peterfabian committed:
Added back check for empty REQUEST_URI to authentication function, as it's still needed, the value is used further down.

claudiosanches committed:
Merge pull request #22764 from woocommerce/renovate/eslint-5.x

Update dependency eslint to v5.14.1

claudiosanches committed:
Merge pull request #22766 from woocommerce/renovate/mocha-6.x

Update dependency mocha to v6

haszari committed:
add header & footer to products empty suggestions box

haszari committed:
style products empty header/footer + support link style:
- added link-text to trigger link style for suggestion CTA, button is now opt-in
- made suggestion rendering more generic, should display something sensible if there are missing attribs
- all CTA links/buttons open in new tab via target attr

haszari committed:
add linkout icon to link-style suggestion CTAs (for empty products footer link)

haszari committed:
allow dismiss by default + hide products empty-state box if no suggestions

haszari committed:
mobile styles for products empty state suggestions

haszari committed:
fix button regression in product table suggestion

haszari committed:
share styles and markup for prpducts list suggestion

mikejolley committed:
Merge pull request #22763 from woocommerce/fix/paypal-bn-code

Move PayPal BN partner ID

mikejolley committed:
Merge pull request #22692 from woocommerce/fix/22686

Update CA address format

mikejolley committed:
Merge pull request #22736 from woocommerce/fix/22703

Importer -  Variations cannot be drafts - set to private.

mikejolley committed:
Merge pull request #22737 from woocommerce/fix/22733

Fix save logic by moving setter into loop grabbing the value

mikejolley committed:
Merge pull request #22740 from woocommerce/fix/22403

Check for decoded taxonomy name when unsetting terms

mikejolley committed:
Merge pull request #22748 from yratof/yratof-patch-price-class

Hard coded class now filterable

mikejolley committed:
Merge pull request #22735 from woocommerce/fix/maybe_set_date_paid

If date paid is not set but payment complete status has passed, set it

mikejolley committed:
Merge pull request #22731 from woocommerce/update/21952

Prevent webhooks for drafts and correct create vs updated for manual orders

mikejolley committed:
Merge pull request #22357 from woocommerce/pr/22125

State and country field handling improvements

mikejolley committed:
Merge pull request #22423 from woocommerce/performance/find-matching-variations

Performance - Finding matching variations

mikejolley committed:
Merge pull request #21090 from woocommerce/fix/20937-v0.1

Added test if request is a REST API request so that cart is not loaded.

mikejolley committed:
Merge pull request #22760 from woocommerce/performance/webhook-loading

Performance: webhook loading

mikejolley committed:
Merge pull request #22329 from woocommerce/enhancement-manual-stock-changes/21754

Product stock adjustments with manual order line item changes

Luminus committed:
Rename Macedonia to North Macedonia

mikejolley committed:
SCSS mailchimp color change missing

mikejolley committed:
Merge pull request #22769 from Luminus/fix/22768

Rename Macedonia to North Macedonia

rrennick committed:
check low_stock for empty string vs isset

rrennick committed:
phpcs except for ignore use of $_POST

rrennick committed:
fixes from review

haszari commented:
Stylelint is complaining about `@include`ing a mixin (`iconbeforedashicons` from `mixins.scss`), how do I address this? (I've been committing this file with `--no-verify` to work around this.)

```
✖ stylelint --syntax=scss --fix found some errors. Please fix them and try committing again.
 
assets/css/marketplace-suggestions.scss
24:2   ✖  Unexpected unknown at-rule "@include"                   at-rule-no-unknown
```

I suspect the linter is confused, since the mixin works correctly when I build styles using `node_modules/grunt/bin/grunt css`.

haszari committed:
hide inline products list suggestion on mobile (can't get the styling right)

haszari committed:
display icon images in suggestions + styling tweaks for header/footer

haszari commented:
Still need to get my js file minified and built – I had problems with the build, so have been developing with `SCRIPT_DEBUG=true`.

renovate-bot committed:
Update dependency autoprefixer to v9.4.8

haszari committed:
fix size of products empty state icon in IE - use pixel sizing

haszari commented:
I've made some changes to `admin.scss` and have been skipping the linter for this too – seeing many errors (20k!), e.g. :

```
✖ stylelint --syntax=scss --fix found some errors. Please fix them and try committing again.
 
assets/css/admin.scss
  20:2   ✖  Unexpected unknown at-rule "@include"                                                                                                                                                                                                                                                                                                                       at-rule-no-unknown                               
 478:5   ✖  Expected selector ".wc_addons_wrap .products li a h3" to come before selector ".wc_addons_wrap .products li a img.extension-thumb + h3"                                                                                                                                                                                                                     no-descending-specificity                        
 499:19  ✖  Unexpected quotes                                                                                                                                                                                                                                                                                                                                           function-url-quotes                              
 506:3   ✖  Expected selector ".wc_addons_wrap .storefront img" to come before selector ".wc_addons_wrap .products li a img"                                         
```

haszari committed:
send tracks events when suggestions are displayed, clicked or dismissed:
- Using temporary stand-in tracks API until PR is merged.
- Also fixed logic so table-view suggestion is not "displayed" if
  empty-state suggestions are shown.

haszari committed:
pull real suggestions data from S3, remove hard-coded test data

mikejolley commented on includes/marketplace-suggestions/class-wc-marketplace-suggestions-api.php:
```diff
@@ -0,0 +1,115 @@
+	}
+
+	$post_data       = wp_unslash( $_POST );
+	$suggestion_slug = sanitize_text_field( $post_data['slug'] );
```
`wc_clean`

haszari replied:
Seems unnecessary, [just a wrapper for `sanitize_text_field`](https://docs.woocommerce.com/wc-apidocs/source-function-wc_clean.html#369-382). Happy to change this if that's preferred around here.

mikejolley commented on templates/marketplace-suggestions/container.php:
```diff
@@ -0,0 +1,17 @@
+<?php
```
This template should not be in the frontend templates directory. We do not want to support theme overrides of this file.

haszari replied:
(now moved)

mikejolley commented on includes/marketplace-suggestions/class-wc-marketplace-suggestions-api.php:
```diff
@@ -0,0 +1,115 @@
+
+	// Parse the data to check for any errors.
+	// If it's valid, store structure in transient.
+	$suggestions = json_decode( wp_remote_retrieve_body( $raw_suggestions ) );
```
Does this needs some kind of sanitization callback on the contents? The JSON is untrusted data.

haszari replied:
Good question – I'm relying on `json_decode` to bail if the data is malformed, and if that succeeds, there's a basic check that it's an array.

Regarding the actual contents, I could add checking for certain keys – I'm not sure how useful that is. Happy to take advice as to the best approach here.

mikejolley commented on assets/js/admin/marketplace-suggestions.js:
```diff
@@ -0,0 +1,340 @@
+			if ( title ) {
+				var titleHeading = document.createElement( 'h4' );
+				titleHeading.textContent = title;
+				left.appendChild( titleHeading );
```
Perhaps some escaping of the data from the JSON should be done here to make sure only text is output and not HTML. There are some workarounds for that like appending the raw content to a div and using `.text()` to grab the safe non-html text.

haszari replied:
We're protected from this because we're assigning `textContent` – any bad stuff gets escaped for us:

<img width="858" alt="screen shot 2019-02-21 at 3 33 40 pm" src="https://user-images.githubusercontent.com/4167300/53139311-1c03a400-35ee-11e9-9734-8fa93f357e08.png">

mikejolley commented on assets/js/admin/marketplace-suggestions.js:
```diff
@@ -0,0 +1,340 @@
+					$( this ).addClass( 'showing-suggestion' );
+					visibleSuggestions.push( promos[i].context );
+
+					window.wcSettings.recordEvent( 'marketplace_suggestion_displayed', {
```
Need to check if this is permitted on .org. Tracks is opt-in but the rule is "Third-party advertisement mechanisms which track usage and/or views." so I don't know if that is an exclusion.

haszari replied:
We're getting this clarified, it's a good point. Will likely follow up separate to this PR.

mikejolley commented on assets/js/admin/marketplace-suggestions.js:
```diff
@@ -0,0 +1,340 @@
+		jQuery.getJSON(
+			ajaxurl,
+			{
+				'action': 'marketplace_suggestions',
```
nonce?

haszari replied:
Since this is just a `GET`, and can't change any server-side state (other than transients/cache), there's no need for a nonce (as far as I'm aware).

mikejolley commented on assets/js/admin/marketplace-suggestions.js:
```diff
@@ -0,0 +1,340 @@
+/* global marketplace_suggestions ajaxurl */
```
You may want to ask someone from Isotope to review this file since I'm more confortable with PHP than JS.

mikejolley replied:
@justinshreve would you be able to take a quick look over this for best practices? tia

mikejolley commented on assets/js/admin/marketplace-suggestions.js:
```diff
@@ -0,0 +1,340 @@
+		function dismissSuggestion( suggestionSlug ) {
+			// hide the suggestion in the UI
+			var selector = '[data-suggestion-slug=' + suggestionSlug + ']';
+			$( selector ).fadeOut();
```
Remove after fade out otherwise the zebra striping is broken once dismissed.

haszari replied:
👍  

https://github.com/woocommerce/woocommerce/pull/22712/commits/e37583bb2b1dc43e54b02ebf52eb34a81f0d709e

mikejolley committed:
Impoved JSON handling

mikejolley commented:
@haszari I forgot to mention in the review, the English wording is showing for any language and isn't translated.

mikejolley committed:
Merge pull request #22777 from woocommerce/fix/22730

check low_stock for empty string vs isset

mikejolley committed:
Merge branch 'master' of https://github.com/woocommerce/woocommerce

claudiosanches committed:
Updated changelog

claudiosanches committed:
Merge pull request #22780 from woocommerce/renovate/autoprefixer-9.x

Update dependency autoprefixer to v9.4.8

haszari commented on includes/admin/list-tables/class-wc-admin-list-table-products.php:
```diff
@@ -47,10 +47,44 @@ public function __construct() {
+	 * Render container for marketplace suggestions.
+	 */
+	protected function render_marketplace_suggestion_empty_state() {
+		wc_get_template(
```
Thanks for the tip – we should make that clear in the docs!

https://docs.woocommerce.com/wc-apidocs/function-wc_get_template.html

haszari replied:
Moved this into the module now, no longer using `wc_get_template`: https://github.com/woocommerce/woocommerce/pull/22712/commits/a9b7bfc513d0da56b7de81d5a717346e1f09e5ec

haszari commented on Gruntfile.js:
```diff
@@ -1,7 +1,7 @@
 module.exports = function( grunt ) {
 	'use strict';
-	const sass = require( 'node-sass' );
+	var sass = require( 'node-sass' );
```
The grunt build was failing, I think a js linter was complaining about this:

```
  Gruntfile.js
      4 |    const sass = require( 'node-sass' );
             ^ 'const' is available in ES6 (use 'esversion: 6') or Mozilla JS extensions (use moz).
```

Since there's no other es6 in this file, or in the rest of the code base (I think!), I thought it was more consistent to keep this simple. I believe it's a recent change, maybe others are working around it differently, or this is fixed upstream now? (I asked about how to address this in proton slack and mentioned my workaround.)

Happy to pull this from this PR, as it's not relevant to what I'm doing. I'm still looking for assistance getting this built/minified correctly though.

kovshenin reviewed (changes requested):
Nice work @haszari! I did a first pass and left some feedback, mainly on the PHP side of things.

My main concern however, is the fact that we're now doing an additional HTTP (AJAX) request on the products page, even if everything has been dismissed and nothing is actually rendered. On some sites with a bunch of slow stuff hooked on `init` this is going to be a pain.

I know I can sometimes be a bit bias towards JavaScript, but in this case we have everything we need in PHP, and we can make a lot of these decisions before rendering the HTML. At the very least I think the contents of the JSON file can be output directly in the HTML, that way we never have to query admin-ajax.php for it, which in this case is essentially just a proxy to `wc_marketplace_suggestions_get_api_data()`.

haszari commented on includes/marketplace-suggestions/class-wc-marketplace-suggestions-api.php:
```diff
@@ -0,0 +1,115 @@
+
+	$raw_suggestions = wp_safe_remote_get(
+		$suggestion_data_url,
+		array( 'user-agent' => 'WooCommerce Marketplace Suggestions' )
```
No particular reason - just following the pattern used for the Featured tab in the addons page:

https://github.com/woocommerce/woocommerce/blob/master/includes/admin/class-wc-admin-addons.php#L27

Happy to remove this.

haszari commented on includes/marketplace-suggestions/class-wc-marketplace-suggestions-api.php:
```diff
@@ -0,0 +1,115 @@
+/**
+ * Return an array of suggestions the user has dismissed.
+ */
+function wc_marketplace_suggestions_get_dismissed() {
```
Correct – that's by design right now.

Note that we should be able set up the json file so the dismiss slug represents a product rather than a particular suggestion instance.

haszari committed:
remove dismissed suggestions from DOM so styling updates accordingly

haszari committed:
cache empty suggestions if API fails, to reduce congestion/thrashing +
+ don't use wp_remote_safe_get – overkill
+ use wp_send_json_success to streamline/simplify API data ajax handler

haszari committed:
remove some extra whitespace

haszari commented on includes/marketplace-suggestions/class-wc-marketplace-suggestions-api.php:
```diff
@@ -0,0 +1,115 @@
+		$suggestion_data_url,
+		array( 'user-agent' => 'WooCommerce Marketplace Suggestions' )
+	);
+	if ( is_wp_error( $raw_suggestions ) ) {
```
Good point – caching it for a day now: https://github.com/woocommerce/woocommerce/pull/22712/commits/dd27bf278b83d34c70a24932563cd1330c6b36f6

haszari commented on includes/marketplace-suggestions/class-wc-marketplace-suggestions-api.php:
```diff
@@ -0,0 +1,115 @@
+function wc_marketplace_suggestions_ajax_handler() {
+	$suggestion_data = wc_marketplace_suggestions_get_api_data();
+
+	echo wp_json_encode( $suggestion_data );
```
Thanks for the tip! The AJAX docs page I was following didn't mention these, I much prefer this! https://github.com/woocommerce/woocommerce/pull/22712/commits/dd27bf278b83d34c70a24932563cd1330c6b36f6

haszari commented on includes/marketplace-suggestions/class-wc-marketplace-suggestions-api.php:
```diff
@@ -0,0 +1,115 @@
+		return $suggestion_data;
+	}
+
+	$raw_suggestions = wp_safe_remote_get(
```
Sounds reasonable - fixed here: https://github.com/woocommerce/woocommerce/pull/22712/commits/dd27bf278b83d34c70a24932563cd1330c6b36f6

haszari committed:
fix client side data access after change to `wp_send_json_success`

haszari committed:
move woo-plugin slugs into a helper routine + cache installed woo plugins in a transient

haszari commented on includes/admin/class-wc-admin-assets.php:
```diff
@@ -416,6 +422,41 @@ public function admin_scripts() {
+		 * @param string $plugin_file Plugin filename.
+		 * @return string
+		 */
+		protected function get_plugin_slug( $plugin_file ) {
```
That makes sense. I've moved this logic and caching into a helper routine: https://github.com/woocommerce/woocommerce/pull/22712/commits/e53f92ae0d53635cc100684fefbfc8da742205e1

mikejolley replied:
Does this method need to be removed then?

haszari commented on includes/admin/class-wc-admin-assets.php:
```diff
@@ -416,6 +422,41 @@ public function admin_scripts() {
+			if ( in_array( $screen_id, array( 'edit-product' ), true ) ) {
+				$woo_plugin_slugs = array_map(
+					array( $this, 'get_plugin_slug' ),
+					array_keys( WC_Helper::get_local_woo_plugins() )
```
Caching for a day: https://github.com/woocommerce/woocommerce/pull/22712/commits/e53f92ae0d53635cc100684fefbfc8da742205e1

haszari committed:
store dismissed suggestions in user meta, don't manually json-encode

haszari commented on includes/marketplace-suggestions/class-wc-marketplace-suggestions-api.php:
```diff
@@ -0,0 +1,115 @@
+		get_current_user_id(),
+		'wc_marketplace_suggestions_dismissed_suggestions',
+		wp_json_encode( $dismissed_suggestions )
+	);
```
TIL about `user_meta` vs `user_option`, agree these dismissals don't need to be per-site.

haszari committed:
tidy up main marketplace suggestions module, make it a real class +
+ require it for admin only
+ fix issue with retrieving dismissed suggestions user meta value

haszari commented on includes/marketplace-suggestions/class-wc-marketplace-suggestions-api.php:
```diff
@@ -0,0 +1,115 @@
+<?php
```
👍 https://github.com/woocommerce/woocommerce/pull/22712/commits/71250c30a2685515150b0c1a28b33bf4fad83868

haszari committed:
move suggestions class into admin/ and require from admin::includes

haszari commented on includes/class-woocommerce.php:
```diff
@@ -378,6 +378,11 @@ public function includes() {
+		/**
+		 * Marketplace suggestions
+		 */
+		include_once WC_ABSPATH . 'includes/marketplace-suggestions/class-wc-marketplace-suggestions-api.php';
```
Good idea, the code and the include are now moved: https://github.com/woocommerce/woocommerce/pull/22712/commits/4c8436122f043f684b8815f3038f89f6b8daf1a6

haszari committed:
move products list suggestions container render into module behind hook + move template near module, don't use wc_get_template to render

haszari commented on includes/admin/list-tables/class-wc-admin-list-table-products.php:
```diff
@@ -47,10 +47,44 @@ public function __construct() {
+
 		echo '</div>';
+
+		$this->render_marketplace_suggestion_empty_state();
```
👍 https://github.com/woocommerce/woocommerce/pull/22712/commits/a9b7bfc513d0da56b7de81d5a717346e1f09e5ec

haszari committed:
support global opt-out of suggestions via user option 'woocommerce_allow_marketplace_suggestions'

haszari commented on includes/admin/class-wc-admin-assets.php:
```diff
@@ -70,6 +72,10 @@ public function admin_styles() {
 				wc_deprecated_function( 'The woocommerce_admin_css action', '2.3', 'admin_enqueue_scripts' );
 			}
+
+			if ( in_array( $screen_id, array( 'edit-product' ), true ) ) {
```
Good idea, added this now: https://github.com/woocommerce/woocommerce/pull/22712/commits/ea5d3870ba9f77f6450391bcbad5c37e1a4b1720

mikejolley replied:
is this missing WC_Marketplace_Suggestions::allow_suggestions()?

haszari committed:
only display suggestions for Englishish locales

haszari committed:
document all js methods

haszari committed:
prefix AJAX actions to reduce chance of collision

haszari commented on includes/marketplace-suggestions/class-wc-marketplace-suggestions-api.php:
```diff
@@ -0,0 +1,115 @@
+ */
+function wc_marketplace_suggestions_api_init() {
+	add_action( 'wp_ajax_marketplace_suggestions', 'wc_marketplace_suggestions_ajax_handler' );
+	add_action( 'wp_ajax_add_dismissed_marketplace_suggestion', 'wc_marketplace_suggestions_dismiss_handler' );
```
I've prefixed the actions, good point. https://github.com/woocommerce/woocommerce/pull/22712/commits/f2deef861e8fcfff20378914a422888863d589b2

This class is now much more self-contained, and the init is triggered from a better place. Let me know if you think we should definitely move the ajax registration somewhere else :)

haszari commented:
> My main concern however, is the fact that we're now doing an additional HTTP (AJAX) request on the products page, even if everything has been dismissed and nothing is actually rendered. On some sites with a bunch of slow stuff hooked on `init` this is going to be a pain.
> In this case we have everything we need in PHP, and we can make a lot of these decisions before rendering the HTML. At the very least I think the contents of the JSON file can be output directly in the HTML.

Interesting comment @kovshenin – initially we were looking at doing more of the work on the server side.

If you feel strongly about this we can take a look at it. Bootstrapping the JSON is definitely doable if that's the better approach.

There are a couple of considerations that let me to this setup:

- Some of these suggestions are much more suited to client side render. Hooking the server-side render to achieve these would have been much uglier/more dangerous code.
- I didn't want to slow down the initial page load; anything we add to the initial request will slow it down and bloat the response (a little bit). I figure we should render the real content as quick as possible; once the user has that and can interact with it, we can then follow up with suggestions in the client side. Since this is 'supplementary' content, I feel it's more appropriate to defer as much of this as we can.

If this is meaningfully impacting the experience of using these pages, then we should look at how to address that – perhaps frontloading would be better.

haszari commented:
Thanks for the reviews @mikejolley & @kovshenin ! I've addressed all the feedback, please take a look and let me know if I've missed anything, or if you spot anything new.

mikejolley commented on assets/js/admin/marketplace-suggestions.js:
```diff
@@ -0,0 +1,360 @@
+		}
+
+		// Render DOM element for suggestion dismiss button.
+		function renderDismissButton( suggestionSlug ) {
```
Not related to this specifically but I noticed the suggestions were dissmissble on blank slate and you can dismiss all instead nothing is left.

Perhaps those ones don't need to be dismissable - they won't be show long anyway, once you've added a product they are forever hidden?

haszari replied:
That's a reasonable suggestion – dismissability is controlled from the JSON data (per suggestion), so we can iterate on that outside of this PR/woo code base.

mikejolley commented on includes/admin/helper/class-wc-helper.php:
```diff
@@ -1046,12 +1046,29 @@ public static function get_local_woo_plugins() {
+	 * (Cached for 1 day.)
+	 */
+	public static function get_local_woo_plugin_slugs() {
+		$woo_plugin_slugs = get_transient( 'wc_helper_installed_woo_plugin_slugs' );
```
I'd actually cache get_local_woo_plugins() instead since that will cover future usage.

As for getting slugs, try `wp_list_pluck` - it's magic.

haszari replied:
Aha! I was googling for a pluck yesterday, but didn't find `wp_list_pluck` or `wc_list_pluck`!

Went with `wp_list_pluck` as we don't need the extra flexibility of the woo wrapper. 48946b3cc6d8e16da067ca35b27ff5d63efbae06

mikejolley commented on includes/admin/marketplace-suggestions/class-wc-marketplace-suggestions.php:
```diff
@@ -0,0 +1,167 @@
+	 */
+	public static function allow_suggestions() {
+		$option = get_option( 'woocommerce_allow_marketplace_suggestions', 'yes' );
+		$allow  = ( 'yes' === $option || 'true' === $option || 1 === $option );
```
We have a helper function for this `wc_string_to_bool`

mikejolley commented on includes/admin/marketplace-suggestions/class-wc-marketplace-suggestions.php:
```diff
@@ -0,0 +1,167 @@
+	 * @return bool
+	 */
+	public static function allow_suggestions() {
+		$option = get_option( 'woocommerce_allow_marketplace_suggestions', 'yes' );
```
Rather than an option this should be a filter because the option has no UI.

e.g. `apply_filters( 'woocommerce_allow_marketplace_suggestions',  true )`

kovshenin replied:
I agree, this should definitely be a filter right before the function returns, so a developer can easily disable it with:

```
add_filter( 'woocommerce_allow_marketplace_suggestions', '__return_false' );
```

haszari replied:
Good points! Addressed in 62d4b5e395893ed1ad35413a7747e8a6eecbfc12

Also added capability check so only users who can install plugins will see suggestions.

One thing I liked about `get_option` was that it could be set via CLI, but filters are probably more well known/understood.

mikejolley commented on includes/admin/marketplace-suggestions/class-wc-marketplace-suggestions.php:
```diff
@@ -0,0 +1,167 @@
+		}
+
+		// Cache empty suggestions data to reduce requests if there are any issues with API.
+		set_transient( 'wc_marketplace_suggestions', '[]', DAY_IN_SECONDS );
```
Should this be a real array perhaps? it's returned on line 142 so should probably not be a string.

haszari replied:
163af89ab3a1cd56bdc862ada6b64086a9db7aa3

mikejolley commented on assets/js/admin/marketplace-suggestions.js:
```diff
@@ -0,0 +1,360 @@
+
+		// Top-level AJAX request to get suggestion data then render suggestions.
+		var data =
+		jQuery.getJSON(
```
Is there any way the JSON could be passed into the script (wp_localize_script) instead of pulled in via AJAX? Ther would be extra overhead doing it this way on all page loads. Even with caching, thats still an extra admin request which would impact performance.

haszari replied:
Now frontloading: a7dd094b3a2e3adb554f0fc8b3f003d34be0dbca

I'll leave you to add a todo or comment or add this to roadmap for `wcSettings`, if appropriate.

kloon commented on includes/admin/marketplace-suggestions/class-wc-marketplace-suggestions.php:
```diff
@@ -0,0 +1,167 @@
+	 * Suggestion data GET handler.
+	 */
+	public static function get_suggestion_json_data_handler() {
+		$suggestion_data = self::get_suggestions_api_data();
```
Instead of calling API when a ajax requests for suggestions are made, why not have a backrground process that runs every x amount of days and then cache the data in the DB and rather look for that option instead? This should solve the issue of failing API requests and visits to certain pages retriggering the requests.

WC_Queue is your friend https://github.com/woocommerce/woocommerce/wiki/WC_Queue---WooCommerce-Worker-Queue

haszari replied:
This sounds perfect! I had trouble getting it to work though – perhaps I'm missing something?

Here's a minimal example (based on the docs):

```php
function my_job_function( $param1, $param2 ) {
    // This is the code that will execute when the job runs.
    error_log( 'queue job running' );

    error_log( 'Param1: ' . $param1 );
    error_log( 'Param2: ' . $param2 );
}
add_action( 'my_job_action', 'my_job_queue', 10, 2 );
error_log( 'registered job' );
WC()->queue()->add( 'my_job_action', array(
	'param1' => 'test',
	'param2' => 'code',
) );
```

When I run this, I get `registered job` in the log, but I haven't managed to confirm the job actually runs. 

I've commited the work-in-progress code to another branch in case we can salvage it. @kovshenin  any ideas?

https://github.com/haszari/woocommerce/commit/4940aad6eaf8d9bb87eda6d271223f14be8483f1

willgorham replied:
You might want `add_action( 'my_job_action', 'my_job_function', 10, 2 );` instead? (replace `my_job_queue`)

haszari replied:
@willgorham dang, good spotting! However, still no luck.

(Edited the code example above after a retest.)

kloon commented on includes/admin/marketplace-suggestions/class-wc-marketplace-suggestions.php:
```diff
@@ -0,0 +1,167 @@
+	 *
+	 * @return array of json API data
+	 */
+	private static function get_suggestions_api_data() {
```
As mentioned previously, rather have this running in a background process at intervals via the WC_Queue system to fetch the data and store it in the DB.

kloon reviewed (changes requested):
A couple of small changes and suggestions, the biggest one would be to offload the API fetching to WC_Queue and have it update at intervals in the background storing the data in the DB/Cache, you can then do all your output and querying via the stored data.

kovshenin commented:
> A couple of small changes and suggestions, the biggest one would be to offload the API fetching to WC_Queue and have it update at intervals in the background storing the data in the DB/Cache, you can then do all your output and querying via the stored data.

@kloon Interesting! I thought about something like this, I was thinking of WP_Cron though, haven't seen WC_Queue before, looks like exactly the thing we need.

The one thing I thought of, however, is that there is a very limited time, between the user installing WooCommerce and adding their first product, so how can we guarantee that the data is actually available, when they first land on the empty products list, as it may be the only time they see the list empty.

If there's a way to make that first call synchronous, I'm fine doing it with AJAX, and from there on have it update the data in the background, and serve via `wp_localize_script()`. Or maybe I'm overthinking it – wouldn't like to over-engineer it either :)

kloon commented:
> If there's a way to make that first call synchronous, I'm fine doing it with AJAX, and from there on have it update the data in the background, and serve via `wp_localize_script()`. Or maybe I'm overthinking it – wouldn't like to over-engineer it either :)

The `add` method on WC_Queue will run the action ASAP, so basically as soon as you call it, it will execute the action. That said, there is a possibility of a delay there due to the nature of Action Scheduler if you have a bunch of queued up actions already running, which should not be the case when you just exit the wizard.

renovate-bot committed:
Update dependency mocha to v6.0.1

kovshenin commented on includes/admin/marketplace-suggestions/class-wc-marketplace-suggestions.php:
```diff
@@ -0,0 +1,167 @@
+	 * @param string $context Suggestion context name (rendered as a css class).
+	 */
+	public static function render_suggestions_container( $context ) {
+		include 'templates/container.php';
```
Relative includes sometimes break on some old and weird PHP setups. Let's try and use absolute includes only.

haszari committed:
frontload suggestion data as script data, remove ajax request

claudiosanches committed:
Merge pull request #22808 from woocommerce/renovate/mocha-6.x

Update dependency mocha to v6.0.1

haszari commented:
@kloon @kovshenin 

I've implemented the frontloading of the API data as script data (https://github.com/woocommerce/woocommerce/pull/22712/commits/a7dd094b3a2e3adb554f0fc8b3f003d34be0dbca), but couldn't get `WC_Queue` working to periodically update. 

This is working well now, and removes the additional AJAX request on page load. The API data is refreshed weekly (or if transient is cleared), so this feels reasonable to me.

Let me know if there are specific improvements we need to make to get this launched, or if you can find the flaw in my `WC_Queue` code!

haszari committed:
use updated tracks API wcSettings.recordEvent => wcTracks.recordEvent

haszari committed:
hide empty-state suggestions for mobile too – focus on desktop for v1

haszari committed:
only enqueue marketplace suggestions styles if they are enabled

haszari committed:
removed redundant `get_plugin_slug`

haszari committed:
Cache local woo plugins info rather than slugs-only subset:
- cache woo_plugins object for more general use
- streamline get_local_woo_plugin_slugs using pluck

haszari committed:
Improve clarity and logic for disabling suggestions:

- Use a filter rather than option
- Use return-early style in function for clarity
- Ensure user has 'install_plugins' capability

haszari committed:
MIA suggestions data should be cached as a real array, not text

haszari commented on assets/js/admin/marketplace-suggestions.js:
```diff
@@ -0,0 +1,360 @@
+
+		// Stand-in wcSettings.recordEvent for until tracks PR is merged.
+		// https://github.com/woocommerce/woocommerce/pull/22746
+		window.wcSettings = window.wcSettings || {};
```
That's the purpose of this line – if wcSettings is undefined, we set it to `{}`. In fact, right now all this logic is firing and protecting us, as we don't yet have any tracks support available in this branch!

haszari replied:
(Note that this has been changed to `wcTracks`, tracks has moved.)

haszari commented on assets/js/admin/marketplace-suggestions.js:
```diff
@@ -0,0 +1,360 @@
+
+						visibleSuggestions.push( context );
+
+						window.wcSettings.recordEvent( 'marketplace_suggestion_displayed', {
```
We define a placeholder method that does nothing to handle this. This approach is pretty common for front-end libs/third-party scripts etc.

haszari committed:
JS code-style fixes

haszari commented on assets/js/admin/marketplace-suggestions.js:
```diff
@@ -0,0 +1,360 @@
+			);
+
+			window.wcSettings.recordEvent( 'marketplace_suggestion_dismissed', {
+				suggestionSlug: suggestionSlug
```
I believe this is only valid in ES6, and I'm not sure that we support that yet. I haven't seen other ES6 features used in woo core code, and I [hit a snag early on](https://github.com/woocommerce/woocommerce/pull/22712/commits/8fc29aaa28bb831ef3f6f5d571942414689004fb) with ES6 `const` breaking the jshint build.

https://ariya.io/2013/02/es6-and-object-literal-property-value-shorthand

justinshreve replied:
Good call - you can ignore this for now 👍.

haszari commented on assets/js/admin/marketplace-suggestions.js:
```diff
@@ -0,0 +1,360 @@
+
+					// dismiss is allowed by default
+					var allowDismiss = true;
+					if ( promos[ 0 ]['allow-dismiss'] === false ) {
```
We need the linter to catch all these!! Didn't realise how many I'd missed!

All these are fixed in https://github.com/woocommerce/woocommerce/pull/22712/commits/79e7699ccda5dca29255597f9ba19a1e5fc576ce - good spotting 🙌

justinshreve replied:
It looks like there is recent-ish linter committed to core, but I'm not sure if it follows all the rules we follow in `wc-admin` yet. Agreed we should make sure its catching everything, and we should make it run automatically on commit :).

haszari committed:
Merge branch master into feature/marketplace-suggestions

# Conflicts:
#assets/css/admin.scss

haszari committed:
use full path when including container template, rename folder 'views'

haszari committed:
allow suggestions to target multiple contexts:

If an array of context strings is specified in json, the suggestion will
be available for display in all specified contexts.

haszari committed:
allow suggestions in multiple contexts:

- fix regression when hiding footer
- rename `visibleSuggestions` for clarity

haszari committed:
limit maximum of 5 suggestions displayed in any given context

haszari commented:
Closing this PR – apologies for the confusion. 

This code is present in another PR: https://github.com/woocommerce/woocommerce/pull/22857

Let's close this PR and wrap up the in-progress reviews in https://github.com/woocommerce/woocommerce/pull/22857. 

cc @kloon @kovshenin @mikejolley @timmyc
